File: tests/e2e/dot_product.py
# RUN: uv run xdsl-exo -o %t.mlir %s
# RUN: mlir-opt --convert-arith-to-llvm --convert-cf-to-llvm --convert-func-to-llvm --reconcile-unrealized-casts %t.mlir | mlir-translate --mlir-to-llvmir | llc -filetype=obj -o %t.o
# RUN: clang %t.o %S/dot_product.c -o %t.bin -lm
# RUN: %t.bin | filecheck %s

from __future__ import annotations

from exo import *


# CHECK: 70.0
@proc
def dot_product(N: size, out: f32[1] @ DRAM, a: f32[N] @ DRAM, b: f32[N] @ DRAM):
    for i in seq(0, N):
        out[0] += a[i] * b[i]

File: tests/e2e/dot_product.c
#include <stdio.h>

extern void dot_product(long N, float *out, float *a, float *b);

int main() {
    float a[4] = {1.0f, 2.0f, 3.0f, 4.0f};
    float b[4] = {5.0f, 6.0f, 7.0f, 8.0f};
    float out[1] = {0.0f};
    dot_product(4, out, a, b);
    printf("%.1f\n", out[0]);
    return 0;
}
